src/chapters/04-mastering-ai-generation.md

```diff
@@ -0,0 +1,534 @@
+g, err := genkit.Init(ctx, 
+    genkit.WithPromptDir("prompts"),
+    genkit.WithPlugins(&googlegenai.GoogleAI{}))
+
+// Load and execute
+recipePrompt := genkit.LookupPrompt(g, "recipe_generator")
+resp, err := recipePrompt.Execute(ctx,
+    ai.WithInput(map[string]any{
+        "cuisine": "Italian", 
+        "servings": 4,
+    }))
+```
+
+### Advanced Features
+
+#### Multi-Message Prompts
+
+Use `{{role}}` for conversations:
+
+```yaml
+---
+model: googleai/gemini-2.5-flash
+input:
+  schema:
+    topic: string
+---
+{{role "system"}}
+You are a helpful cooking instructor.
+
+{{role "user"}}
+Explain {{topic}} in simple terms.
+```
+
+#### Multi-Modal Support
+
+Handle images with `{{media}}`:
+
+```yaml
+---
+model: googleai/gemini-2.5-flash
+input:
+  schema:
+    imageUrl: string
+---
+{{role "user"}}
+What dish is shown in this image?
+{{media url=imageUrl}}
+```
+
+#### Partials for Reusability
+
+Create reusable components:
+
+```yaml
+# prompts/_base_personality.prompt
+You are a helpful assistant specialized in {{domain}}.
+Always be concise and accurate.
+```
+
+Use in other prompts:
+
+```yaml
+---
+model: googleai/gemini-2.5-flash
+---
+{{role "system"}}
+{{>base_personality domain="cooking"}}
+
+{{role "user"}}
+{{question}}
+```
+
+#### Custom Helpers
+
+Register helpers for custom logic:
+
+```go
+genkit.DefineHelper(g, "shout", func(input string) string {
+    return strings.ToUpper(input)
+})
+```
+
+Use in prompts:
+
+```yaml
+---
+model: googleai/gemini-2.5-flash
+input:
+  schema:
+    name: string
+---
+
+HELLO, {{shout name}}!!!
+```
+
+#### Prompt Variants
+
+Test different versions:
+
+```text
+prompts/
+├── analyzer.prompt          # Baseline
+├── analyzer.detailed.prompt # Variant A
+└── analyzer.concise.prompt  # Variant B
+```
+
+Load dynamically:
+
+```go
+variant := "detailed" // from A/B test
+prompt := genkit.LookupPrompt(g, "analyzer." + variant)
+```
+
+> **Note**: For production A/B testing, you can integrate with [Firebase Remote Config](https://firebase.google.com/docs/remote-config) or [Firestore](https://firebase.google.com/docs/firestore) to dynamically control variant selection and enable gradual rollouts.
+
+### Developer UI Integration
+
+![](../images/test-dotprompt.png)
+
+The Developer UI provides a visual interface for experimenting with prompts before committing them to code. To use it effectively, your application needs to stay running:
+
+```go
+func main() {
+    ctx := context.Background()
+
+    // Initialize Genkit with prompt directory
+    _, err := genkit.Init(ctx,
+        genkit.WithPlugins(&googlegenai.GoogleAI{}),
+        genkit.WithPromptDir("prompts"),
+    )
+    if err != nil {
+        log.Fatalf("could not initialize Genkit: %v", err)
+    }
+
+    // Block execution to keep the app running for Developer UI
+    // This allows genkit start to connect to your application
+    // Even without flows, the Developer UI needs the app running to test prompts
+    select {}
```

Comment:
maybe expose the go app as a server instead of this pattern?

Reply:
In addition to the golang server, I also implemented middleware: [cc0116f](https://github.com/mastering-genkit/mastering-genkit-go/pull/6/commits/cc0116f355a88130ba7326331891cecd5f9e0e6b)